oneoffs/opening_freqs_export.py

```diff
@@ -190,39 +192,106 @@ def create_hourly_reports(hour_directory):
+
+
+def run_time_ranges(db):
+    ts = lambda hr: int(dt.datetime.strptime(hr, "%Y-%m-%d-%H").timestamp())
+    runs = [ (r[0], (ts(r[1]), ts(r[2]))) for r in db.execute('''
+        select run, min(hour), max(hour) from joseki_counts group by 1;
+        ''').fetchall()]
+    return dict(runs)
+
+
+def build_run_time_transformers(ranges):
+    """ Build a dict of functions to transform from a timestamp into a relative
+    offset.  E.g.
+    input: {'v17': (1234567890, 1235567879) ... }
+    output: {'v17': lambda t: (t - min) * (1/max) ... }
+    """
+
+    funcs = {}
+    def f(t, min_, max_):
+        #return "%0.2f" % ((t-min_) * (1/(max_-min_)))
+        key = (t-min_) * (1/(max_-min_))
+        return "%0.3f" % (int(250*key) / 2.5)
+
+    for run, range_ in ranges.items():
+        funcs[run] = functools.partial(f, min_=range_[0], max_=range_[1])
+
+    return funcs
+
+
 def create_top_report():
     """
     Creates an html page showing the most common sequences in the database, and
     charting their popularity over time.
     """
+    db = sqlite3.connect(FLAGS.db_path)
     ts = lambda hr: int(dt.datetime.strptime(hr, "%Y-%m-%d-%H").timestamp())
     print('querying')
-    db = sqlite3.connect(FLAGS.db_path)
-    cur = db.execute('''
-                     select seq, sum(count) from joseki_counts group by seq order by 2 desc limit 300;
-                     ''')
-    sequences = [row for row in cur.fetchall()]
-
-    sequences = collapse_common_prefix(sequences)
-    sequences.sort(key=lambda s: s[1], reverse=True)
-    print("Pruned to ", len(sequences))
-
-    data = []
-    for seq, count in sequences[:30]:
-        print(seq, count)
-        cur = db.execute('''
-                         SELECT hour, count from joseki_counts where seq = ?;
-                         ''', (seq,))
-
-        dat = cur.fetchall()
-        dat = [(ts(d[0]), d[1]) for d in dat]
-        obj = {"count": count, "chart": dat, "sequence": seq}
-        data.append(obj)
-
-    print('saving')
-    tmpl = Template(open('oneoffs/joseki.html').read())
-    with open(FLAGS.out_file, 'w') as out:
-        out.write(tmpl.render(giant_blob=json.dumps(data)))
+    ranges = run_time_ranges(db)
+    interps = build_run_time_transformers(ranges)
+    seqs_by_run = top_seqs_by_run(db, 100)
+    runs = sorted(seqs_by_run.keys())
+
+    cols = []
+    cols.append({'id': 'time', 'label': '% of Training', 'type': 'number'})
+    for run in runs:
+        cols.append({'id': run + 'count', 'label': run + ' times seen', 'type': 'number'})
+
+    for run in runs:
+        data = []
+        sequences = seqs_by_run[run]
+        for seq, count in sequences:
+            print(run, seq, count)
+            rows = collections.defaultdict(lambda: [0 for i in range(len(runs))])
+
+            for idx, r in enumerate(runs):
+                cur = db.execute('''
+                                 SELECT hour, count, run from joseki_counts where seq = ? and run = ?;
```

Comment:
silly to return run when you know what is it.

SELECT hour, count from joseki_counts where seq = ? and run = ?;

Reply:
Done.